### PR TITLE
feat: Implementing DB-API types according to the PEP-0249 specification

### PR DESCRIPTION
**Pull request by mf2199**
The proposed changes represent reworked `types.py` in accordance with the **PEP-0249** [Type Objects and Constructors](https://www.python.org/dev/peps/pep-0249/#type-objects-and-constructors) specifications.

_Some of the classes implemented prior to the alpha release are not part of the DB-API standards. Those were left in for now but may be removed later in the GA release._

### CODE REVIEW

File: google/cloud/spanner_dbapi/types.py
Comment (IlyaFaer) [inline google/cloud/spanner_dbapi/types.py:19]: I think we should first add system tests for these types, before merging. As they are gonna be used to convert data from user types to the Spanner types, we better check how it's working in the case close to the real.

Comment (mf2199) [reply google/cloud/spanner_dbapi/types.py]: @IlyaFaer Reverted to the previous convention, PTAL.

**Review by skuruppu — APPROVED** [review]
I think this change is reasonable and looks a lot cleaner now. Just to have another pair of eyes, I'll add @larkee as a reviewer but if this is urgent, feel free to merge.

**Pull request merged**
